### PR TITLE
✨(backends) es granular backend

### PR DESCRIPTION
**Pull request by SergioSim**
## Purpose

An extract from the `more granular backends` PR #464 concerning only the ESDataBackend

## Proposal

- [x] Adapt the ESDataBackend to use the template method pattern

### CODE REVIEW

File: src/ralph/backends/data/base.py [hunk 1]
```diff
@@ -228,6 +296,47 @@ def write(  # pylint: disable=too-many-arguments
...
+
+        data = iter(data)
+        try:
+            first_record = next(data)
+        except StopIteration:
+            self.logger.info("Data Iterator is empty; skipping write to target.")
+            return 0
+        data = chain((first_record,), data)
+
+        chunk_size = chunk_size if chunk_size else self.settings.DEFAULT_CHUNK_SIZE
```
Comment (Leobouloc) [inline src/ralph/backends/data/base.py:314]: not linked to this PR in particular but perhaps we could have `DEFAULT_READ_CHUNK_SIZE` & `DEFAULT_WRITE_CHUNK_SIZE` .

File: src/ralph/backends/data/base.py [hunk 2]
```diff
@@ -79,24 +116,20 @@ def wrapper(*args, **kwargs):
...
-
-        Args:
-            settings (BaseDataBackendSettings or None): The data backend settings.
-                If `settings` is `None`, a default settings instance is used instead.
-        """
-
+    # This method is replaced by validate_backend_query
     def validate_query(
         self, query: Union[str, dict, BaseQuery, None] = None
     ) -> BaseQuery:
```
Comment (Leobouloc) [inline src/ralph/backends/data/base.py:135]: So the intention is to remove this method ?

**Review by Leobouloc — COMMENTED** [review]
Read up to utils.py. Looks cool, this will make things neater !

File: src/ralph/backends/base.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        """Pydantic Configuration."""
+
+        env_prefix = "RALPH_BACKENDS__"
+
+
+class BaseBackend(ABC):
+    """Base ralph backend class."""
+
+    settings_class = BaseBackendSettings
+    settings: settings_class
```
Comment (quitterie-lcs) [inline src/ralph/backends/base.py:31]: I don't understand why you have to do this ? Why not defining one attribute ?

File: src/ralph/backends/data/base.py (hunk 1 above)
Comment (quitterie-lcs) [reply src/ralph/backends/data/base.py:314]: I think it is a good idea

Comment (SergioSim) [reply src/ralph/backends/data/base.py:314]: Great Idea) I think we could do this change easily once all backends use this pattern)

File: src/ralph/backends/data/base.py (hunk 2 above)
Comment (SergioSim) [reply src/ralph/backends/data/base.py:135]: Yes) partially) we move it out of the class, add two new arguments to it (query_class and logger) and rename it)

File: src/ralph/backends/base.py (hunk 3 above)
Comment (SergioSim) [reply src/ralph/backends/base.py:31]: Well spotted) We do not have to do it) However, static-type checkers (e.g. pylance) seem to struggle to infer the type of the settings attribute(
With the generic version, this issue is not present anymore.

**Comment by SergioSim** [comment]
This PR was the first attempt to shorten PR #464 by splitting it by backend.
However, for more simplicity, we opted to split the PR #464 by functionality in PRs: #501 #510 #511 #512 #513 #515 #517 #518 #519

**Pull request closed without merging**